design/service-binding.md

```diff
@@ -0,0 +1,870 @@
+
+Now, the Service Binding Operator can discover the bootstrap information using the following status descriptor in the `ClusterServiceVersion`:
+
+``` yaml
+apiVersion: operators.coreos/com:v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: strimzi-cluster-operator.v0.16.2
+  namespace: placeholder
+  annotations:
+    containerImage: docker.io/strimzi/operator:0.16.2
+  categories:
+  - Bindable
+spec:
+  customresourcedefinitions:
+    owned:
+    - description: Represents a Kafka cluster
+      displayName: Kafka
+      kind: Kafka
+      name: kafkas.kafka.strimzi.io
+      version: v1beta1
+      statusDescriptors:
+      - description: The addresses of the bootstrap servers for Kafka clients
+        displayName: Bootstrap servers
+        path: listeners
+        x-descriptors:
+        - servicebinding:endpoints:elementType=sliceOfMaps,sourceKey=type,sourceValue=bootstrap
+```
+
+#### Consuming client's ServiceBinding
+
+All of the required annotations are applied to the `Kafka` CSV, so the binding should only need to refer to the `Kafka` CR.
+
+``` yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBinding
+metadata:
+  name: my-binding
+spec:
+  services:
+  - group: kafka.strimzi.io
+    kind: Kafka
+    version: v1beta1
+    resourceRef: my-cluster
+```
+
+The consuming client needs to know the listener type.
+
+The Service Binding Operator creates binding information which contains:
+
+* **endpoints.<listener_name>** - bootstrap server information for all listeners
+
+This information will be provided to the consuming client as an environment variable unless overridden.
+
+
+## Binding to a Kafka cluster with TLS but no authentication
+
+The addition with this scenario is that Kafka clients need access to the CA certificate that signed the broker's server certificate. The Service Binding Specification does not currently have support for a separate CA certificate, which seems like a simple enhancement, which it indicated using `caSecret` in the example below.
+
+The CA certificate is most easily obtained from the `<cluster>-cluster-ca-cert` secret. While this name is predictable, it is not known to the Service Binding Operator. To enable the Service Binding Operator to obtain this information, the same pattern of enhancing the CR `status` and annotating the CSV can be used.
+
+``` yaml
+apiVersion:kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+ ...
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: my-cluster-kafka-bootstrap.my-namespace.svc
+      port: 9092
+    bootstrap: my-cluster-kafka-bootstrap.my-namespace.svc:9092
+  - type: tls
+    addresses:
+    - host: my-cluster-kafka-bootstrap.my-namespace.svc
+      port: 9093
+    bootstrap: my-cluster-kafka-bootstrap.my-namespace.svc:9093
+  caCertificateSecret: my-cluster-cluster-ca-cert
```

Comment:
Also, should this be _per listener_?

Reply:
Thanks for the review. I think I have a little more work to do on the certificates part of this.

All of the CSV stuff can also be achieve with annotations.

What I propose to do then is start to deliver this in pieces:

I'll try to get the first part into Strimzi 0.18:
- Augment the Kafka CR status to include a string of bootstrap servers. Once this is done, an adventurous user wanting to use SBO can easily get the list of bootstrap servers without having to use a Go template.
- Potentially put the matching CSV x-descriptor in also.

Then Strimzi 0.18 would be able to support basic binding scenarios.

After 0.18, I propose:
- Continue working on the certificates section. I've clearly not quite got it right and I need to have a solution that can encompass all of the advanced uses of certificates with Strimzi, while also being easily consumable by the end user. This last part is important.
- Deliver the remaining code once we're all agreed on how to handle certificates, as well as the remaining CSV x-descriptors.
- Keep in step with the Service Binding Operator as it evolves. Specifically, it's important to keep on top of how information is merged into secrets and that is still a bit fluid right now.
- Enhance the document to include the use of annotations as well as CSV x-descriptors.
- Document how to use service binding with Strimzi.

Reply:
Sounds good to me ... but it would be great to get some feedback also from other @strimzi/maintainers ... not just from me.

Reply:
This sounds like a very reasonable plan forward.
For future reference, is the intention that @AndrewJSchofield edit the csv directly in https://github.com/operator-framework/community-operators
Or have we got a separate repo that we make changes to prior to pushing to the community?

Reply:
The CSV lives in the Operator Framework repo and we manually update it after each release. We used to have it locally, but it was useless since at the end you still need to go there.

After the release we can work with @AndrewJSchofield to get the first changes there when we prepare the 0.18 CSV.